pkg/controller/controller.go

```diff
@@ -0,0 +1,96 @@
+package controller
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"strings"
+	"time"
+
+	"github.com/rancher/k3k/pkg/apis/k3k.io/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	namePrefix      = "k3k"
+	k3SImageName    = "rancher/k3s"
+	AdminCommonName = "system:admin"
+)
+
+// Backoff is the cluster creation duration backoff
+var Backoff = wait.Backoff{
+	Steps:    5,
+	Duration: 5 * time.Second,
+	Factor:   2,
+	Jitter:   0.1,
+}
+
+func K3SImage(cluster *v1alpha1.Cluster) string {
+	return k3SImageName + ":" + cluster.Spec.Version
+}
+
+func LogAndReturnErr(errString string, err error) error {
+	klog.Errorf("%s: %v", errString, err)
+	return err
+}
+
+func nodeAddress(node *v1.Node) string {
+	var externalIP string
+	var internalIP string
+
+	for _, ip := range node.Status.Addresses {
+		if ip.Type == "ExternalIP" && ip.Address != "" {
+			externalIP = ip.Address
+			break
+		}
+		if ip.Type == "InternalIP" && ip.Address != "" {
+			internalIP = ip.Address
+		}
+	}
+	if externalIP != "" {
+		return externalIP
+	}
+
+	return internalIP
+}
+
+// return all the nodes external addresses, if not found then return internal addresses
+func Addresses(ctx context.Context, client ctrlruntimeclient.Client) ([]string, error) {
+	var nodeList v1.NodeList
+	if err := client.List(ctx, &nodeList); err != nil {
+		return nil, err
+	}
+
```

Comment:
Should we have a len check on the list to if it's > 1, perform the action below?

Reply:
I dont think so, because we wont need the node addresses if the nodes arent registered yet